all-sdk/external-client-configuration.md

```diff
@@ -0,0 +1,340 @@
+* Well-specified hierarchy for on-disk overridable by environment variables overridable by in-CLI/SDK values
+* 0 new dependencies in SDKs
+* Profile names within the configuration
+
+### Non-goals
+
+* Common format for declarative workers, schedules, etc. This is only for client options.
+* File-based hierarchy. We don't need many levels of file merging for these options at this time.
+* Extreme focus on human authoring of this format. This is because it'd likely go against the "0 new dependencies" rule.
+* Use external configuration by default. Users need to opt-in to this for compatibility and clarity reasons (but it is
+  of course easy).
+
+## Specification
+
+### Values and File Format
+
+* The format is a JSON object with the following fields:
+  * `profiles` - Object with keys as profile names. The default profile should be named `default`.
+    * `<name>` - Profile object with configuration values. Profile names are case-insensitive. 💭 Why? They are needed
+      for environment variables which are case insensitive. It is a validation error to have a config file with two
+      separately-cased profile names that are equal case-insensitively.
+      * `address` - Client address, aka gRPC "host:port".
+      * `namespace` - Client namespace.
+      * `apiKey` - Client API key.
+      * `tls` - A boolean (true is same as empty object) _or_ an object. Note the default for this value is dependent
+        upon the client and may change. 💭 Why? We regret not making TLS the default in the past. If TLS is an object,
+        it can have the following possible fields:
+        * `clientCertPath` - File path to mTLS client cert. Mutually exclusive with `clientCertData`.
+        * `clientCertData` - Cert data. Mutually exclusive with `clientCertPath`.
+        * `clientKeyPath` - File path to mTLS client key. Mutually exclusive with `clientKeyData`.
+        * `clientKeyData` - Key data. Mutually exclusive with `clientKeyPath`.
+        * `serverCaCertPath` - File path to server CA cert. Mutually exclusive with `serverCaCertData`.
+        * `serverCaCertData` - CA cert data. Mutually exclusive with `serverCaCertPath`.
+        * `serverName` - Override SNI name when connecting to server.
+        * `disableHostVerification` - Boolean to disable verifying TLS server host. May not be available in all Temporal
+          clients. 💭 Why? Not all SDKs offer this feature today.
+      * `codec` - Remote codec information to use for all encoding/decoding of payloads. May not be available in all
+        Temporal clients. 💭 Why? Not all SDKs allow use of a remote codec today.
+        * `endpoint` - Endpoint to the remote codec.
+        * `auth` - Authorization header for the remote codec.
+      * `grpcMeta` - Object representing HTTP headers. Values must be strings for now (can discuss arrays later).
+        ❓ Should we call this `headers`? We do in some cases and not others.
```

Comment:
:shrug: I kinda lean towards headers since it's maybe slightly more what people are used to, but, meta is more accurate

Reply:
I am leaning there as well, but I would need to suggest a CLI update for that.